contracts/utils/cryptography/P256.sol

```diff
@@ -0,0 +1,316 @@
+     */
+    function isOnCurve(uint256 x, uint256 y) internal pure returns (bool result) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            let p := P
+            let lhs := mulmod(y, y, p)
+            let rhs := addmod(mulmod(addmod(mulmod(x, x, p), A, p), x, p), B, p)
+            result := eq(lhs, rhs)
+        }
+    }
+
+    /**
+     * @dev Reduce from jacobian to affine coordinates
+     * @param jx - jacobian coordinate x
+     * @param jy - jacobian coordinate y
+     * @param jz - jacobian coordinate z
+     * @return ax - affine coordinate x
+     * @return ay - affine coordinate y
+     */
+    function _affineFromJacobian(uint256 jx, uint256 jy, uint256 jz) private view returns (uint256 ax, uint256 ay) {
+        if (jz == 0) return (0, 0);
+        uint256 zinv = _invModP(jz);
+        uint256 zzinv = mulmod(zinv, zinv, P);
+        uint256 zzzinv = mulmod(zzinv, zinv, P);
+        ax = mulmod(jx, zzinv, P);
+        ay = mulmod(jy, zzzinv, P);
+    }
+
+    /**
+     * @dev Point addition on the jacobian coordinates
+     * https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates
+     */
+    function _jAdd(
+        uint256 x1,
+        uint256 y1,
+        uint256 z1,
+        uint256 x2,
+        uint256 y2,
+        uint256 z2
+    ) private pure returns (uint256 x3, uint256 y3, uint256 z3) {
+        if (z1 == 0) {
+            return (x2, y2, z2);
+        }
+        if (z2 == 0) {
+            return (x1, y1, z1);
+        }
+        /// @solidity memory-safe-assembly
+        assembly {
+            let p := P
+            let zz1 := mulmod(z1, z1, p) // zz1 = z1²
+            let zz2 := mulmod(z2, z2, p) // zz2 = z2²
+            let u1 := mulmod(x1, zz2, p) // u1 = x1*z2²
+            let u2 := mulmod(x2, zz1, p) // u2 = x2*z1²
+            let s1 := mulmod(y1, mulmod(zz2, z2, p), p) // s1 = y1*z2³
+            let s2 := mulmod(y2, mulmod(zz1, z1, p), p) // s2 = y2*z1³
+            let h := addmod(u2, sub(p, u1), p) // h = u2-u1
+            let hh := mulmod(h, h, p) // h²
+            let hhh := mulmod(h, hh, p) // h³
+            let r := addmod(s2, sub(p, s1), p) // r = s2-s1
+
+            // x' = r²-h³-2*u1*h²
+            x3 := addmod(addmod(mulmod(r, r, p), sub(p, hhh), p), sub(p, mulmod(2, mulmod(u1, hh, p), p)), p)
+            // y' = r*(u1*h²-x')-s1*h³
+            y3 := addmod(mulmod(r, addmod(mulmod(u1, hh, p), sub(p, x3), p), p), sub(p, mulmod(s1, hhh, p)), p)
+            // z' = h*z1*z2
+            z3 := mulmod(h, mulmod(z1, z2, p), p)
+        }
+    }
+
+    /**
+     * @dev Point doubling on the jacobian coordinates
+     * https://en.wikibooks.org/wiki/Cryptography/Prime_Curve/Jacobian_Coordinates
+     */
+    function _jDouble(uint256 x, uint256 y, uint256 z) private pure returns (uint256 x2, uint256 y2, uint256 z2) {
+        /// @solidity memory-safe-assembly
+        assembly {
+            let p := P
+            let yy := mulmod(y, y, p)
```

Comment:
Shouldn't we return the point at infinity (0,1,0) if `y == 0` at this point?

Reply:
This part should be discussed over a call, but here are my first thought

A jacobian point (x,y,z) represents the carthesian point (x/z², y/z³). Therefore, any jacobian point that has `z=0` is at infinity (because x/z² and y/z³ are x/0 and y/0, which is not defined).

For the case of `_jDouble`, if y=0 we have `z' = 2*y*z = 0`

So my understanding is that, if y=0 we know the result is a point at infinity, and we can potentially skip some computation is we already know one. However, if we don't skip, and we do the computation "normaly", we get a point at infinity, which is correct. We could have skipped the computation, but given that in our case `addmod` and `mulmod` are "cheap native operation" (which is not the case in an x86 machine), its probably ok.

We can try implementing the skip, and see if its actually saving gas, but AFAIK, this is an optimisation issue and not a correctness issue.

Reply:
Ok, I was missing why we're using these coordinates in the first place. I found [this article](https://hackmd.io/@cailynyongyong/HkuoMtz6o) that helped me understand better. 

So if I'm understanding correctly, we prefer operating in the jacobian space because it has advantages over the affine space. The transformation from affine coordinates to jacobian corresponds to what you described above, but the point with `z = 0` is just undefined given the representation (i.e. `x/0**2` or `y/z**3`)

In such case then it doesn't matter to return here because the returned point will also be undefined (`z' = 0`).
I'm solving now it makes sense, please point me out if I'm getting something wrong